src/ledger/LedgerManagerImpl.cpp

```diff
@@ -1782,6 +2023,397 @@ LedgerManagerImpl::prefetchTransactionData(AbstractLedgerTxnParent& ltx,
+                                           ThreadEntry{nullopt, true});
+                }
+            }
+        }
+    }
+}
+
+void
+LedgerManagerImpl::writeGlobalEntryMapToLedgerTxn(
+    AbstractLedgerTxn& ltx, ThreadEntryMap const& globalEntryMap)
+{
+    LedgerTxn ltxInner(ltx);
+    for (auto const& entry : globalEntryMap)
+    {
+        // Only update if dirty bit is set
+        if (!entry.second.isDirty)
+        {
+            continue;
+        }
+
+        if (entry.second.mLedgerEntry)
+        {
+            auto const& updatedEntry = *entry.second.mLedgerEntry;
+            auto ltxe = ltxInner.load(entry.first);
+            if (ltxe)
+            {
+                ltxe.current() = updatedEntry;
+            }
+            else
+            {
+                ltxInner.create(updatedEntry);
+            }
+        }
+        else
+        {
+            auto ltxe = ltxInner.load(entry.first);
+            if (ltxe)
+            {
+                ltxInner.erase(entry.first);
+            }
+        }
+    }
+    ltxInner.commit();
+}
+
+void
+LedgerManagerImpl::applySorobanStage(AppConnector& app, AbstractLedgerTxn& ltx,
+                                     ThreadEntryMap& globalEntryMap,
+                                     ApplyStage const& stage,
+                                     Hash const& sorobanBasePrngSeed)
+{
+    auto const& config = app.getConfig();
+    auto const& sorobanConfig = getSorobanNetworkConfigForApply();
+    auto ledgerInfo = getParallelLedgerInfo(app, ltx.loadHeader().current());
+
+    auto [threadRestoredKeys, entryMapsByCluster] =
+        applySorobanStageClustersInParallel(app, ltx, stage, globalEntryMap,
+                                            sorobanBasePrngSeed, config,
+                                            sorobanConfig, ledgerInfo);
+
+    addAllRestoredKeysToLedgerTxn(threadRestoredKeys, ltx);
+
+    checkAllTxBundleInvariantsAndCallProcessPostApply(app, stage, config,
+                                                      ledgerInfo, ltx);
+
+    auto isInReadWriteSet = getReadWriteKeysForStage(stage);
+
+    writeDirtyMapEntriesToGlobalEntryMap(entryMapsByCluster, globalEntryMap,
+                                         isInReadWriteSet);
+}
+
+void
+LedgerManagerImpl::applySorobanStages(AppConnector& app, AbstractLedgerTxn& ltx,
+                                      std::vector<ApplyStage> const& stages,
+                                      Hash const& sorobanBasePrngSeed)
+{
+    ThreadEntryMap globalEntryMap;
+    // From now on, we will be using globalEntryMap, liveSnapshots, and the
+    // hotArchive to collect all entries. Before we continue though, we need to
+    // load the fee source accounts because they have already been updated with
+    // charged fees, and preParallelApply will update sequence numbers which
+    // won't be reflected in the live snapshot.
+    preParallelApplyAndCollectFeeSourceAccounts(app, ltx, stages,
```

Comment:
This is a subtle change from the previous iteration. We now call `preParallelApply` on all txs at once, instead of calling it prior to every stage.